src/box/lua/connpool.lua

```diff
@@ -271,46 +329,111 @@ local function filter(opts)
         end
     end
+
     -- Return if retrieving dynamic information is not required.
-    if next(static_candidates) == nil or next(dynamic_opts) == nil then
+    if mode == 'any' then
         return static_candidates
     end
 
     -- Filter the remaining candidates after connecting to them.
-    local connected_candidates = connect_to_candidates(static_candidates)
+    local connected_instances = acquire_all_candidates(static_candidates)
+
     local dynamic_candidates = {}
-    for _, instance_name in pairs(connected_candidates) do
+    for _, instance_name in pairs(connected_instances) do
         if is_candidate_match_dynamic(instance_name, dynamic_opts) then
             table.insert(dynamic_candidates, instance_name)
         end
     end
+
     return dynamic_candidates
 end
 
+-- Subroutine for the `call()` function. Tries to find an active
+-- connection matching the requirements and considering the
+-- specified priorities.
 local function get_connection(opts)
-    local mode = nil
+    assert(opts.mode == 'alive' or opts.mode == 'ro' or opts.mode == 'rw' or
+           opts.mode == 'prefer_ro' or opts.mode == 'prefer_rw')
+
+    local mode = 'alive'
     if opts.mode == 'ro' or opts.mode == 'rw' then
         mode = opts.mode
     end
-    local candidates_opts = {
+
+    -- It's better to use the local instance to perform the call
+    -- faster. So prefer the local instance by default.
+    local prefer_local = true
+    if opts.prefer_local ~= nil then
+        prefer_local = opts.prefer_local
+    end
+
+    local dynamic_opts = {
+        mode = mode,
+    }
+
+    local local_instance_name = box.info.name
+
+    local filter_static_opts = {
         groups = opts.groups,
         replicasets = opts.replicasets,
         instances = opts.instances,
         labels = opts.labels,
         roles = opts.roles,
         sharding_roles = opts.sharding_roles,
-        mode = mode,
     }
-    local candidates = filter(candidates_opts)
+    local static_candidates = filter(filter_static_opts)
+
+    -- In case we don't need to prioritize the matching instances
+    -- we can wait for any connection available.
+    if opts.mode ~= 'prefer_rw' and opts.mode ~= 'prefer_ro' then
+        if prefer_local then
+            for _, instance_name in ipairs(static_candidates) do
+                if instance_name ~= local_instance_name then
+                    goto continue
+                end
+
+                -- This connect is almomst always quick and successful,
+                -- so it's being waited. Though it can fail e.g. due to
+                -- the FD limit. That's why it still should be checked.
+                local ok, conn = pcall(connect, local_instance_name,
+                                       { wait_connected = true })
+
+                if not ok then break end
+
+                if is_candidate_match_dynamic(local_instance_name,
+                                              dynamic_opts) then
```

Comment:
AFAIR, `wait_connected = true` doesn't guarantee that `:mode()` already has ro/rw state to filter on it. So the `is_candidate_match_dynamic` may return a wrong result (false negative).

It doesn't look as a problem, in fact (we'll just use another connection), but at least worth to comment.

However, I would better skip the connection explicitly if the mode is not received yet.

Reply:
I've added the explicit check and the comment explaining the problem. Thank you!

Reply:
Now this code looks so:

```lua
-- To use the local connection, the connect() call
-- should be ok, if the mode is specified the
-- connection mode should exist (thus, this select
-- may be false-negative) and the candidate should
-- match dynamic options.
if ok and (opts.mode == nil or conn:mode() ~= nil) and
   is_candidate_match_dynamic(local_instance, opts) then
    return conn
end
```

To be honest, I wouldn't understand what you want to say here: `conn:mode()` may be `nil` during some period after `connect` above, so `is_candidate_match_dynamic` may discard the connection.